### PR TITLE
Nightshade Typescript refactoring

### LINKED ISSUE #5: NEARLIB: no explicit error if trying to sign transaction, but no key is found.

### LINKED ISSUE #12: bs58 should be smarter
```
    /**
     * Encode a buffer as string using bs58
     * @param {Buffer} buffer 
     * @example
     * KeyPair.encodeBufferInBs58(key.publicKey)
     */
    static encodeBufferInBs58(buffer) {
        const bytes = Buffer.from(buffer);
        const encodedValue = bs58.encode(bytes);
        return encodedValue;
    }
```
Stuff like this should be inside the bs58.encode() function instead of outside, bs58 should auto determine the type is Buffer or Uint8Array.  Maybe use a forked bs58 with this update?

Or get a base58 lib that supports uint8array buffers.

### LINKED ISSUE #18: Better error reporting
Hard to interpret error messages like:
> Error: [-32603] Internal error: Tx (00) not found

### LINKED ISSUE #27: Clean-up 'make function calls' test in integration.test.js
@vgrichina commented on [Fri Feb 15 2019](https://github.com/nearprotocol/nearcore/issues/579)

Test is currently whole wall of code with few unrelated tests jammed together:

```js
    test('make function calls', async () => {
        const args = {
            'name': 'trex'
        };
        const viewFunctionResult = await nearjs.callViewFunction(
            contractName,
            'hello', // this is the function defined in hello.wasm file that we are calling
            args);
        expect(viewFunctionResult).toEqual('hello trex');

        var setCallValue = await generateUniqueString('setCallPrefix');
        const setArgs = {
            'value': setCallValue
        };
        const scheduleResult = await nearjs.scheduleFunctionCall(
            0,
            aliceAccountName,
            contractName,
            'setValue', // this is the function defined in hello.wasm file that we are calling
            setArgs);
        expect(scheduleResult.hash).not.toBeFalsy();
        await nearjs.waitForTransactionResult(scheduleResult);
        const getValueResult = await nearjs.callViewFunction(
            contractName,
            'getValue', // this is the function defined in hello.wasm file that we are calling
            {});
        expect(getValueResult).toEqual(setCallValue);

        // test that load contract works and we can make calls afterwards
        const options = {
            viewMethods: ['hello', 'getValue'],
            changeMethods: ['setValue'],
            sender: aliceAccountName
        };
        const contract = await nearjs.loadContract(contractName, options);
        const helloResult = await contract.hello(args);
        expect(helloResult).toEqual('hello trex');
        var setCallValue2 = await generateUniqueString('setCallPrefix');
        const setArgs2 = {
            'value': setCallValue2
        };
        const setValueResult = await contract.setValue(setArgs2);
        expect(setValueResult.status).toBe('Completed');
        const getValueResult2 = await contract.getValue();
        expect(getValueResult2).toEqual(setCallValue2);
    });
```

compare to other tests, e.g.:

```js
    test('can get logs from method result', async () => {
        await contract.generateLogs();
        expect(logs).toEqual([`[${contractName}]: LOG: log1`, `[${contractName}]: LOG: log2`]);
    });
```

---

@ilblackdragon commented on [Tue Mar 26 2019](https://github.com/nearprotocol/nearcore/issues/579#issuecomment-476840403)

@janedegtiareva any update here?

### PR DESCRIPTION
**Pull request by ilblackdragon**
This follows ideas for #39 + TypeScript for better typed APIs.

Fixes #5 
Fixes #12 
Ref #13 
Fixes #18 
Fixes #27

### CODE REVIEW

File: test/account.test.js [hunk 1]
```diff
@@ -28,12 +34,29 @@ test('create account and then view account returns the created account', async (
...
+    expect(state.amount).toEqual(INITIAL_BALANCE + BigInt(10000));
+});
+
 // Generate some unique string with a given prefix using the alice nonce.
 const generateUniqueString = async (prefix) => {
     let status = await masterAccount.state();
     return prefix + status.nonce;
 };
+
+async function testCreateAccount() {
```
Comment (vgrichina) [inline test/account.test.js]: 1) `s/testCreateAccount/createAccount/`
2) I think there should be ability to override default values (like amount) to make it more flexible, i.e. similar to this https://github.com/nearprotocol/nearlib/blob/master/test/integration.test.js#L175

File: src.ts/transaction.ts [hunk 2]
```diff
@@ -1,3 +1,6 @@
+'use strict';
+
+import { BN } from "bn.js";
```
Comment (vgrichina) [inline src.ts/transaction.ts]: I think we shouldn't use `bn.js`, but use https://github.com/GoogleChromeLabs/jsbi which is forward-compatible with native JS BigInt.

File: test/account.test.js (hunk 1 above)
Comment (vgrichina) [reply test/account.test.js]: Might also be worth looking into some of the test object factory modules:
https://www.npmtrends.com/factory-girl-vs-factory-js-vs-fixture-factory-vs-rosie

File: src.ts/transaction.ts (hunk 2 above)
Comment (ilblackdragon) [reply src.ts/transaction.ts]: added comment explaining why we use it.

File: test/account.test.js (hunk 1 above)
Comment (ilblackdragon) [reply test/account.test.js]: Hm, not sure about the factory stuff - if it's test fixture stuff can do separately.

**Pull request merged**